### PR TITLE
db:  add flow nums to task jobs table

### LINKED ISSUE #5252: Update task_jobs table in database
Add `flow_nums` to the `task_jobs` table. `flow_nums` are currently held in `task_states` table. 

This is required to solve the current problem with cylc review - detailed in https://github.com/cylc/cylc-flow/issues/5247

Requires a database upgrader.

### PR DESCRIPTION
**Pull request by wxtim**
Closes #5252 

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [ ] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by datamel** [comment]
Has something weird happened to the rebase? I'm seeing another pr's changes in here.

File: cylc/flow/task_events_mgr.py [hunk 1]
```diff
@@ -1277,6 +1278,7 @@ def _insert_task_job(
                 'submit_status': submit_status,
                 'time_submit_exit': event_time,
                 'job_id': itask.summary.get('submit_method_id'),
+                'flow_nums': serialise(itask.flow_nums),
```
Comment (oliver-sanders) [inline cylc/flow/task_events_mgr.py:1281]: I'm not sure we need to set this here. I think `put_update_task_jobs` should be called only with fields which have changed. The `flow_nums` field should be set when the job is added to the DB but never changed.

I think it's the `put_insert_task_jobs` calls which need this information.

**Comment by wxtim** [comment]
> Has something weird happened to the rebase? I'm seeing another pr's changes in here.

Yes - I forgot to change this to being a PR against 8.0.x rather than master.

File: CHANGES.md [hunk 2]
```diff
@@ -33,6 +33,11 @@ command to validate, install and play a workflow.
 [#5081](https://github.com/cylc/cylc-flow/pull/5081) - Reduced amount that
 gets logged at "INFO" level in scheduler logs.
 
+### Fixes
+
```
Comment (oliver-sanders) [inline CHANGES.md:37]: Doesn't really fix anything in of itself, perhaps list as an enhancement.

```suggestion
```

Comment (oliver-sanders) [reply CHANGES.md:37]: Consider adding a breaking changes section for this:

```md
# Breaking Changes

* Workflows started with Cylc 8.0 which contain multiple "flows" cannot be restarted with
  Cylc 8.1 due to DB changes.
```

File: tests/unit/test_workflow_db_mgr.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        r"    ('foo', '10050101T0000Z', '[1]',"
+        r" '2022-12-05T14:46:33Z',"
+        r" '2022-12-05T14:46:40Z', 1, 'succeeded', 0, 0)"
+    )
+    db_file_name = _setup_db(values)
+    pri_dao = CylcWorkflowDAO(db_file_name)
+    WorkflowDatabaseManager.upgrade_pre_804(pri_dao)
+    conn = sqlite3.connect(db_file_name)
+    result = conn.execute('SELECT flow_nums FROM task_jobs;').fetchall()[0][0]
+    assert result == '[1]'
```
Comment (oliver-sanders) [inline tests/unit/test_workflow_db_mgr.py:92]: Consider doing a `SELECT DISTINCT` to make sure all rows have been set correctly.

**Review by oliver-sanders — COMMENTED** [review]
LGTM

**Comment by oliver-sanders** [comment]
```
FAILED tests/integration/test_workflow_db_mgr.py::test_db_upgrade_pre_803 - sqlite3.OperationalError: duplicate column name: flow_nums
```

**Review by oliver-sanders — APPROVED** [review]
Ignore me, being daft.

The upgrader logic currently *cannot* handle upgrades from earlier in the same dev version. So long as it works correctly with official releases all good.

**Review by datamel — APPROVED** [review]
Checked out and tested. Code also read. No problems found from me. Thanks @wxtim.

**Pull request merged**